### DISCUSSION TITLE
Overlay plot with multiple value fields of a domain object

### DISCUSSION THREAD

**Discussion by Duncan-McD** [nasa/openmct#3883] [Community Q&A] [answered]
Hello,

For my research team I am trying to implement a vector grapher to graph each coordinate in an overlay. The way it is set up currently is that the telem object is called vectorA and it has child values of raw_vectorA, x_vectorA, y_vectorA, z_vectorA

Is there a way to include a duplicate version of the same telem object on an overlay or stacked plot so I could view the x,y,and z values of the same field on the same plot? currently what I was able to determine is that if you drag in the telemetry field you can select which of the 4 values you want to graph (so i could choose to graph x_vectorA for instance) but if I then tried to re-drag in the same telem object into the stacked or overlay plot to then change that to y_vectorA, it doesn't let you drag it in. Does anyone know a work around for this or if using overlays in this way is achievable?

Thanks,

Duncan

**Comment by Duncan-McD** [comment]
Here is an example of the type of object we want to access and plot values from:

![vector_example](https://user-images.githubusercontent.com/71865023/116790281-48929080-aa81-11eb-82fd-0693a151cb29.png)

**Comment by jvigliotta** [comment, marked as answer]
Open MCT does have the ability to duplicate objects, if you right click on the item in the object tree on the left, you should have the option to "Duplicate". Let me know if that's what you're looking for if that helps you out!

**Reply by nunoguedelha** [reply]
Hello @jvigliotta ,

thank you for the support. Unfortunately it seems that in this case, the "Duplicate" option is available only for the "Overlay Plot" container itself, not the plots dragged into it.

I was able to achieve this "duplication" by doing as follows:
1. Export the Overlay Plot to a JSON file
2. Edit the JSON file, duplicating the linked domain object in the "composition" and "configuration" arrays and adding a "yKey" field for each.
3. Remove the Overlay Plot from the workspace.
4. Import the modified JSON file (this recreates the modified Overlay Plot)

For @Duncan-McD 's use case, the modified JSON file would be:
```
{
  "openmct": {
    "<some alpha-numerical code>": {
      "composition": [
        {
          "namespace": "<vectorA namespace>",
          "key": "vectorA"
        },
        {
          "namespace": "<vectorA namespace>",
          "key": "vectorA"
        },
        {
          "namespace": "<vectorA namespace>",
          "key": "vectorA"
        },
        {
          "namespace": "<vectorA namespace>",
          "key": "vectorA"
        }
      ],
      "configuration": {
        "series": [
          {
            "identifier": {
              "namespace": "<vectorA namespace>",
              "key": "vectorA"
            },
            "yKey": "raw_vectorA"
          },
          {
            "identifier": {
              "namespace": "<vectorA namespace>",
              "key": "vectorA"
            },
            "yKey": "x_vectorA"
          },
          {
            "identifier": {
              "namespace": "<vectorA namespace>",
              "key": "vectorA"
            },
            "yKey": "y_vectorA"
          },
          {
            "identifier": {
              "namespace": "<vectorA namespace>",
              "key": "vectorA"
            },
            "yKey": "z_vectorA"
          }
        ],
        "yAxis": {
          "label": "vectorA raw / x / y / z"
        },
        "xAxis": {}
      },
      "name": "vectorA Overlay Plot",
...
```

But after I imported the JSON file, the plotted field was reset to the same for both plots in the Overlay Plot. I modified it manually through the GUI, and it worked.

The problem is that as soon as I click elsewhere and come back and click on the the Overlay Plot again, the configuration was reset and the same field was selected in both plots. Probably, when the Overlay Plot function initializes the plots on the click event, it only sees one object and not two distinct.

### Here is an illustration of my use case:

<img width="1089" alt="image" src="https://user-images.githubusercontent.com/6848872/124302027-e88f9780-db60-11eb-9d21-9cd0827060ea.png">

- "IMU sensor measurements" is the original domain object.
- "value.ori.roll" and "value.ori.pitch" are the fields I want to display in the same Overlay Plot.

@jvigliotta , would you have an idea on how to avoid the problem I'm having?

@Duncan-McD , any news on your side?

**Reply by Duncan-McD** [reply]
Hello @nunoguedelha ,

I had tried the JSON approach as well a while back and ended with the same thing as you (it would switch back to the same fields when returning to that object).

I had a limited amount of time and wasn't getting anywhere with the front end, so I ended up splitting the subfields of a domain object I wanted to plot on an overlay into individual domain objects (in my telemetry provider), which wasn't the cleanest solution but was what worked best in the time frame I had.

**Reply by nunoguedelha** [reply]
Thanks for the information @Duncan-McD . I might have to adopt the same workaround you mentioned, but I still hope there is a proper solution for this. Let's see ...

**Comment by jvigliotta** [comment]
Hey all, thanks for the details. Currently, there is no native way to handle this without the workarounds you've mentioned, but @shefalijoshi is working on something that may be helpful in this situation. Once she's implemented the improvement, we'll post an update here.

**Reply by nunoguedelha** [reply]
Thanks @jvigliotta , that's really good news! Would you have a rough idea of when @shefalijoshi's change would be available if it is a short development?

I ask this to see if I can afford waiting for it and avoid the workaround. We have pretty big vectors to plot in our project.

**Reply by jvigliotta** [reply]
Hey @nunoguedelha, I'll get with @shefalijoshi to see if she has a timeframe and some general details of what the fix would be.

**Reply by nunoguedelha** [reply]
Thanks a lot @jvigliotta !

**Comment by shefalijoshi** [comment]
@nunoguedelha I don't have a timeframe at this point, however, I'll be sure to update this issue when I have it.

**Reply by nunoguedelha** [reply]
Thank you @shefalijoshi . Looking forward to see your updates.

**Reply by nunoguedelha** [reply]
Hello @shefalijoshi , @jvigliotta , any news on this?
